### PR TITLE
[web] Enforce onDrawFrame/onBeginFrame render rule

### PR DESCRIPTION
**Pull request by harryterkelsen**
Also update tests to conform to the render rule.

The render rule is that the call to `render` should only occur in the scope of an `onDrawFrame` or `onBeginFrame` call. As a side effect, this change also fixes a memory leak which occurs in tests using FakeAsync in the `flutter_test` environment (because they don't render anymore).

Fixes https://github.com/flutter/flutter/issues/137073
Fixes https://github.com/flutter/flutter/issues/137669

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/platform_dispatcher.dart [hunk 1]
```diff
@@ -80,7 +80,7 @@ abstract class PlatformDispatcher {
 
   void scheduleFrame();
 
-  void render(Scene scene, [FlutterView view]);
+  Future<void> render(Scene scene, [FlutterView view]);
```
Comment (yjbanov) [inline lib/web_ui/lib/platform_dispatcher.dart:83]: It looks like the mobile version no longer has this method.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart [hunk 2]
```diff
@@ -753,14 +761,23 @@ class EnginePlatformDispatcher extends ui.PlatformDispatcher {
...
-    renderer.renderScene(scene, view ?? implicitView!);
+    final ui.FlutterView viewToRender = view ?? implicitView!;
+
+    // Only render in an `onDrawFrame` or `onBeginFrame` scope. This is checked
+    // by checking if the `_viewsRenderedInCurrentFrame` is non-null and this
+    // view hasn't been rendered already in this scope.
+    final bool shouldRender =
+        _viewsRenderedInCurrentFrame?.add(viewToRender) ?? false;
+    if (shouldRender) {
+      await renderer.renderScene(scene, view ?? implicitView!);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:779]: nit: s/`view ?? implicitView!`/`viewToRender`/

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart (hunk 2 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:779]: done

File: lib/web_ui/lib/platform_dispatcher.dart (hunk 1 above)
Comment (ditman) [reply lib/web_ui/lib/platform_dispatcher.dart:83]: Agh shoot, forgot to remove it from our interface 🤦

**Pull request merged**